Adding support for MAX17043 Lipo battery-gauge

Vincent1964 opened the pull request:
## Description:

This PR contains a sensor driver for MAX17043 a battery gauge for 3.7 Volt batteries. Allows to monitor a battery powered ESP.

**Related issue (if applicable):** fixes - NONE

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [x] Only relevant files were touched
  - [x] Only one feature/fix was added per PR and the code change compiles without warnings
  - [ ] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [x] The code change is tested and works with Tasmota core ESP32 V.2.0.9
  - [x] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

Vincent1964 committed:
Inital version, ready for debugging session.

Vincent1964 committed:
update settings

Vincent1964 committed:
Adding missing comma after function pointer

Vincent1964 committed:
update

Vincent1964 committed:
First working version, mqqt testing pending

Vincent1964 committed:
update defines in core

Vincent1964 committed:
Finalizing, final touch. Ready for PR

Vincent1964 committed:
Fixing typo

Vincent1964 commented:
First PR, hope I did everything the way it should be.

Vincent1964 reviewed (commented):
already fixed by: [18737](https://github.com/arendst/Tasmota/discussions/18737)

arendst commented on .vscode/settings.json:
Pls remove this file from your PR.

Vincent1964 replied:
done

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,136 @@
+#ifdef USE_WEBSERVER
+void Max17043Show(void) {
+
+  WSContentSend_P(PSTR("Battery: %.1f V,  "), max17043->voltage/1000);
```
- Use standard Tasmota GUI layout (Search code for examples)
- Use standard Tasmota ``%1_f`` instead of not supported (because adds loads of unneeded code size) ``%.1f``

Vincent1964 replied:
fixed

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,136 @@
+void Max17043Show(void) {
+
+  WSContentSend_P(PSTR("Battery: %.1f V,  "), max17043->voltage/1000);
+  WSContentSend_P(PSTR("%.1f %%"), max17043->percentage);
```
Ditto. Search for ``HTTP_BATTERY`` for an example.

Vincent1964 replied:
fixed

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,136 @@
+void Max17043Json(void) {
+
+  ResponseAppend_P(JSON_SNS_BGAUGE, mqttId, String(max17043->voltage/1000, 2), String(max17043->percentage, 1));
+}
```
Don't use ``String()`` as in takes way too much code (size). See other drivers how to use ``%2_f``.

Vincent1964 replied:
fixed

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,136 @@
+\*********************************************************************************************/
+
+bool Xsns109(uint32_t function) {
+
```
Add chk for I2cEnabled(). See other I2c driver how to use.

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,136 @@
+    AddLog(LOG_LEVEL_ERROR, PSTR("I2C: MAX17043 not found"));
+  } else {
+    max17043 = (MAX17043 *)calloc(1, sizeof(struct MAX17043));
+    AddLog(LOG_LEVEL_INFO, PSTR("I2C: MAX17043 initialized"));
```
Remove AddLog and add I2cSetActiveFound() claiming it's I2C address.

Your driver needs to integrate in Tasmota's I2C eco-system.

Vincent1964 replied:
fixed

Vincent1964 committed:
Delete settings.json

Vincent1964 committed:
solving issues

arendst commented on tasmota/tasmota_xx2c_global/xx2c_interface.ino:
No need to change anything in this file. Remove from your PR

Vincent1964 commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,136 @@
+
+void Max17043Init(void) {
+
+  if (gauge.begin() != 0) {
```
fixed

Vincent1964 commented on tasmota/tasmota_xx2c_global/xsns_interface.ino:
ok, will be removed

Vincent1964 committed:
Fixed all comments from Arendst

Vincent1964 committed:
Delete xsns_interface.ino

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,130 @@
+
+#ifdef USE_WEBSERVER
+void Max17043Show(void) {
+  WSContentSend_PD(PSTR("{s}%s Voltage  {m}%1_f V  {e}"), SENSOR_NAME, &max17043->voltage);
```
Pls chg "Voltage" with defined D_VOLTAGE making it multi language.

Add "Capacity" as english word to all language files; translations will eventually be done by others.

arendst commented on tasmota/tasmota_xx2c_global/xsns_interface.ino:
By removing this file the PR fails initial chks. Pls find another solution to fix PR and making it ready to merge.

Vincent1964 committed:
adding i18n strings and undelete file

Vincent1964 committed:
modified:   tasmota/language/af_AF.h
modified:   tasmota/language/bg_BG.h
modified:   tasmota/language/ca_AD.h
modified:   tasmota/language/cs_CZ.h
modified:   tasmota/language/de_DE.h
modified:   tasmota/language/el_GR.h
modified:   tasmota/language/en_GB.h
modified:   tasmota/language/es_ES.h
modified:   tasmota/language/fr_FR.h
modified:   tasmota/language/fy_NL.h
modified:   tasmota/language/he_HE.h
modified:   tasmota/language/hu_HU.h
modified:   tasmota/language/it_IT.h
modified:   tasmota/language/ko_KO.h
modified:   tasmota/language/nl_NL.h
modified:   tasmota/language/pl_PL.h
modified:   tasmota/language/pt_BR.h
modified:   tasmota/language/pt_PT.h
modified:   tasmota/language/ro_RO.h
modified:   tasmota/language/ru_RU.h
modified:   tasmota/language/sk_SK.h
modified:   tasmota/language/sv_SE.h
modified:   tasmota/language/tr_TR.h
modified:   tasmota/language/uk_UA.h
modified:   tasmota/language/vi_VN.h
modified:   tasmota/language/zh_CN.h
modified:   tasmota/language/zh_TW.h
modified:   tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,130 @@
+
+#ifdef USE_WEBSERVER
+void Max17043Show(void) {
+  WSContentSend_PD(PSTR("{s}%s D_VOLTAGE {m}%1_f V  {e}"), SENSOR_NAME, &max17043->voltage);
```
Getting close! Look how they did it here: ```"{s}%s " D_VOLTAGE "{m}%1_f " D_UNIT_VOLT "{e}"```. Notice the quotes.

arendst commented on tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino:
```diff
@@ -0,0 +1,130 @@
+#ifdef USE_WEBSERVER
+void Max17043Show(void) {
+  WSContentSend_PD(PSTR("{s}%s D_VOLTAGE {m}%1_f V  {e}"), SENSOR_NAME, &max17043->voltage);
+  WSContentSend_PD(PSTR("{s}%s D_BATTERY_CAPACITY {m}%1_f %% {e}"), SENSOR_NAME, &max17043->percentage);
```
Ditto quotes.

Once changed pls test your code as this error should have been noticed by you too.

Vincent1964 replied:
I think it worked, I normaly test before PR. Anyway, thanks for your patience with me!!

Jason2866 commented:
`.vscode/settings.json` is still deleted with your PR, and you are doing changes in files not related with your PR. Please update your branch with latest development and remove the not needed changes in files.

Vincent1964 committed:
modified:   .gitignore

Vincent1964 committed:
modified:   .gitignore

Vincent1964 commented:
To do that I have to revoke my PR, update and make a new PR, right?


Thanks for your patience!

Vincent1964 committed:
modified:   tasmota/tasmota_xsns_sensor/xsns_109_max17043.ino

Jason2866 commented:
Probably easiest is to close this PR and open a new one.

Vincent1964 commented:
For me yes but not for the Theo and other reviewers they then have to review all my code again.. I will try to remove the file from the PR

arendst commented:
I agree with jason. 

Take your time anyway as I won’t be able to merge the next fifteen days.

Vincent1964 commented:
Okay Theo, relax
